elks/arch/i86/drivers/char/console-direct.c

```diff
@@ -41,34 +49,86 @@
 
 #define MAXPARMS        28
 
+#define N_DEVICETYPES   2
+#define MAX_OUTPUTS     2
+
+struct output;
+typedef struct output Output;
 struct console;
 typedef struct console Console;
 
+enum OutputType {
+    OT_MDA = 0,
+    OT_CGA,
+    OT_EGA,
+    OT_VGA,
+};
+
 struct console {
+    unsigned char op;
     int cx, cy;                 /* cursor position */
     void (*fsm)(Console *, int);
     unsigned char attr;         /* current attribute */
     unsigned char XN;           /* delayed newline on column 80 */
-    unsigned int vseg;          /* video segment for page */
-    int basepage;               /* start of video ram */
+    unsigned int vseg;          /* vram for this console page */
+    int vseg_offset;            /* vram offset of vseg for this console page */
 #ifdef CONFIG_EMUL_ANSI
     int savex, savey;           /* saved cursor position */
     unsigned char *parmptr;     /* ptr to params */
     unsigned char params[MAXPARMS];     /* ANSI params */
 #endif
+    enum OutputType type;
+    unsigned short w;
+    unsigned short max_col;
+    unsigned short h;
+    unsigned short max_row;
+    unsigned short crtc_base;   /* 6845 CRTC base I/O address */
+};
+
+struct hw_params {
+    unsigned short w;
+    unsigned short h;
+    unsigned short crtc_base;
+    unsigned vseg_base;
+    int vseg_bytes;
+    unsigned char init_bytes[16];
+    unsigned char n_init_bytes;
+    unsigned char max_pages;
+    unsigned short page_words;
+    unsigned char is_present;
 };
```

Comment:
I'm leaning towards the idea of getting all this hardware-specific structures and code moved out of console-direct.c into a standalone monitor-ibmpc.{c,h} (or something like that). Then call the code as functions when gated.

This will remove quite a bit of added complexity from an already complicated console-direct.c. What do you think?
That also allows you to gets as complicated as needed within that file and have it remain hidden from review by those curious about general console operation.

Reply:
Yeah, I think a file like `6845.c` at least would make sense. And I think some of this probing/setup code could also be marked as `INITPROC`, so we free it after booting, right?

Reply:
> I think a file like 6845.c

Good idea: crtc-6845.c would match the naming scheme we have in elks/arch/i86/kernel/timer-8254.c etc.

And yes on INITPROC as well, as console_init is already INITPROC. The biggest issue we have with far functions is that the compiler has to generate thunking code when those functions call near functions, which adds time and space. But in the case of INITPROC far functions, that extra code, if any, doesn't matter as much because it is released after boot. Calling far procs from near or far procs don't need thunking code.

Reply:
Is there a INITPROC equivalent for data that is only needed at init time?

Reply:
Unfortunately no. I had previously added to my todo list to look at whether GCC can store data in the code segment and then generate CS: overrides to get to it. That INITDATA could then be discarded along with INITPROC code.

We might be able to place the data into its own section with some tricky .section naming and linker scripts, but I still don't know a way to force GCC to reload DS when accessing that data. So basically it'll just be part of the larger implementation with CONFIG_CONSOLE_DUAL.

Reply:
ACK, something for a future PR perhaps!